pkg/controller/controller.go

```diff
@@ -678,3 +735,26 @@ const (
 	reconcileDelete ReconciliationAction = "Delete"
 	reconcilePoll   ReconciliationAction = "Poll"
 )
+
+func (c *controller) getClusterID() (id string) {
+	// Use the read lock for reading, so that multiple instances
+	// provisioning at the same time do not collide.
+	c.clusterIDLock.RLock()
+	id = c.clusterID
+	c.clusterIDLock.RUnlock()
+
+	// fast exit if it exists
+	if id != "" {
+		return
+	}
+	// lazily create on first access if does not exist
```

Comment:
There is a race condition when the goroutine monitoring the configmap and the goroutine reconciling service instances try to get the cluster ID concurrently. They could both pass the first part of this routine, seeing that a cluster ID has not been set, then both attempt to set the cluster ID. This could result in a service instance using the incorrect cluster ID.

I think it would be much simpler to only ever have the cluster ID set by the goroutine that is monitoring the configmap. If the service instance reconciliation fails temporarily because the cluster ID has not been set yet, that is fine. Everything will resolve itself in short order.

Reply:
Code is updated, please look below.